### PR TITLE
timerslack improvements for linux and usermode waiting 

### PR DESCRIPTION
**Pull request by Whatcookie**
Uses user mode wait instructions when available, and an appropriate number of threads on the machine are detected.

Should improve power efficiency on windows particularly for users with Ryzen and at least 12 threads, or users with 12th gen intel and at least 12 threads.

Also has some improvements for linux timers, thanks to the ability to set the timerslack.

### CODE REVIEW

File: rpcs3/Emu/Cell/lv2/lv2.cpp [hunk 1]
```diff
@@ -1880,6 +1890,34 @@ void lv2_obj::set_yield_frequency(u64 freq, u64 max_allowed_tsc)
 	g_lv2_preempts_taken.release(0);
 }
 
+#if defined(_MSC_VER)
+#define mwaitx_func
+#define waitpkg_func
+#endif
```
Comment (Nekotekina) [inline rpcs3/Emu/Cell/lv2/lv2.cpp:1896]: Did you mean #else?

Comment (Whatcookie) [reply rpcs3/Emu/Cell/lv2/lv2.cpp]: Oops, yes

File: rpcs3/main.cpp [hunk 2]
```diff
@@ -1048,6 +1049,10 @@ int main(int argc, char** argv)
 		}
 	}
 
+#ifdef __linux__
+	prctl(PR_SET_TIMERSLACK, 1, 0, 0, 0);
```
Comment (Megamouse) [inline rpcs3/main.cpp:1053]: Missing comment.

Comment (Whatcookie) [reply rpcs3/main.cpp:1054]: Added a comment

File: rpcs3/Emu/Cell/lv2/lv2.cpp [hunk 3]
```diff
@@ -1965,14 +2005,29 @@ bool lv2_obj::wait_timeout(u64 usec, ppu_thread* cpu, bool scale, bool is_usleep
...
+				u32 us_in_tsc_clocks = remaining * (utils::get_tsc_freq() / 1000000);
+
+				if (utils::has_waitpkg())
+				{
+					__tpause(us_in_tsc_clocks, 0x0);
+				}
+				else if (utils::has_waitx())
+				{
+					__mwaitx(us_in_tsc_clocks, 0x00);
+				}
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/lv2.cpp:2028]: Else path does nothing?

File: rpcs3/Emu/Cell/lv2/lv2.cpp [hunk 4]
```diff
@@ -1880,6 +1891,35 @@ void lv2_obj::set_yield_frequency(u64 freq, u64 max_allowed_tsc)
...
+#define mwaitx_func __attribute__((__target__("mwaitx")))
+#define waitpkg_func __attribute__((__target__("waitpkg")))
+#endif
+
+#if defined(ARCH_X64)
+// Waits for a number of TSC clock cycles in power optimized state
+// Cstate is represented in bits [7:4]+1 cstate. So C0 requires bits [7:4] to be set to 0xf, C1 requires bits [7:4] to be set to 0.
+mwaitx_func static void __mwaitx(u32 cycles, u32 cstate)
+{
+	const u32 timerEnable = 0x2;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/lv2.cpp:1907]: constexpr.
snake_case

**Comment by Whatcookie** [comment]
Results from @Yahfz's 13900K on windows:

<Details>
<summary> Before </summary>

image

image


</Details>

<Details>
<summary> After </summary>

image

image

</Details>

In both cases we see higher performance with lower power draw.

File: rpcs3/Emu/Cell/lv2/lv2.cpp [hunk 5]
```diff
@@ -1965,14 +2005,29 @@ bool lv2_obj::wait_timeout(u64 usec, ppu_thread* cpu, bool scale, bool is_usleep
 			if (remaining > host_min_quantum)
 			{
 #ifdef __linux__
-				// Do not wait for the last quantum to avoid loss of accuracy
-				wait_for(remaining - ((remaining % host_min_quantum) + host_min_quantum));
+				// With timerslack set low, Linux is precise for all values above 10us
```
Comment (kd-11) [inline rpcs3/Emu/Cell/lv2/lv2.cpp:2008]: The "hack" on line 1991 is now unnecessary with timerslack. It should just be 10 at all times.

**Review by kd-11 — APPROVED** [review]

**Comment by Nekotekina** [comment]
Needs rebase

**Pull request merged**